crates/fuel-core/src/graphql_api/indexation/balances.rs

```diff
@@ -0,0 +1,598 @@
+    let MessageBalance {
+        mut retryable,
+        mut non_retryable,
+    } = current_balance;
+    if message.is_retryable_message() {
+        retryable = retryable.saturating_add(message.amount() as u128);
+    } else {
+        non_retryable = non_retryable.saturating_add(message.amount() as u128);
+    }
+    let new_balance = MessageBalance {
+        retryable,
+        non_retryable,
+    };
+
+    block_st_transaction
+        .storage::<MessageBalances>()
+        .insert(key, &new_balance)
+        .map_err(Into::into)
+}
+
+fn decrease_message_balance<T>(
+    block_st_transaction: &mut T,
+    message: &Message,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = message.recipient();
+    let storage = block_st_transaction.storage::<MessageBalances>();
+    let MessageBalance {
+        retryable,
+        non_retryable,
+    } = storage.get(key)?.unwrap_or_default().into_owned();
+    let current_balance = if message.is_retryable_message() {
+        retryable
+    } else {
+        non_retryable
+    };
+
+    let new_amount = current_balance
+        .checked_sub(message.amount() as u128)
+        .ok_or_else(|| IndexationError::MessageBalanceWouldUnderflow {
+            owner: *message.recipient(),
+            current_amount: current_balance,
+            requested_deduction: message.amount() as u128,
+            retryable: message.is_retryable_message(),
+        })?;
+
+    let new_balance = if message.is_retryable_message() {
+        MessageBalance {
+            retryable: new_amount,
+            non_retryable,
+        }
+    } else {
+        MessageBalance {
+            retryable,
+            non_retryable: new_amount,
+        }
+    };
+    block_st_transaction
+        .storage::<MessageBalances>()
+        .insert(key, &new_balance)
+        .map_err(Into::into)
+}
+
+fn increase_coin_balance<T>(
+    block_st_transaction: &mut T,
+    coin: &Coin,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = CoinBalancesKey::new(&coin.owner, &coin.asset_id);
+    let storage = block_st_transaction.storage::<CoinBalances>();
+    let current_amount = storage.get(&key)?.unwrap_or_default().into_owned();
+    let new_amount = current_amount.saturating_add(coin.amount as u128);
+
+    block_st_transaction
+        .storage::<CoinBalances>()
+        .insert(&key, &new_amount)
+        .map_err(Into::into)
+}
+
+fn decrease_coin_balance<T>(
+    block_st_transaction: &mut T,
+    coin: &Coin,
+) -> Result<(), IndexationError>
+where
+    T: OffChainDatabaseTransaction,
+{
+    let key = CoinBalancesKey::new(&coin.owner, &coin.asset_id);
+    let storage = block_st_transaction.storage::<CoinBalances>();
+    let current_amount = storage.get(&key)?.unwrap_or_default().into_owned();
+
+    let new_amount =
+        current_amount
+            .checked_sub(coin.amount as u128)
```

Comment:
```suggestion
            .checked_sub(u128::from(coin.amount))
```

Reply:
Updated in 5eb9bb06309ef064a281043fc32a8aaffc916989